lib/Sema/TypeCheckAttr.cpp

```diff
@@ -876,39 +876,41 @@ void AttributeChecker::visitSPIAccessControlAttr(SPIAccessControlAttr *attr) {
     // If VD is a public protocol requirement it can be SPI only if there's
     // a default implementation.
     if (auto protocol = dyn_cast<ProtocolDecl>(D->getDeclContext())) {
-      auto implementations = TypeChecker::lookupMember(
-                                             D->getDeclContext(),
-                                             protocol->getDeclaredType(),
-                                             VD->createNameRef(),
-                                             NameLookupFlags::ProtocolMembers);
-      bool hasDefaultImplementation = llvm::any_of(implementations,
-        [&](const LookupResultEntry &entry) {
-          auto entryDecl = entry.getValueDecl();
-          auto DC = entryDecl->getDeclContext();
-          auto extension = dyn_cast<ExtensionDecl>(DC);
-
-          // The implementation must be defined in the same module in
-          // an unconstrained extension.
-          if (!extension ||
-              extension->getParentModule() != protocol->getParentModule() ||
-              extension->isConstrainedExtension())
-            return false;
+      if (!protocol->isSPI()) {
+        auto implementations = TypeChecker::lookupMember(
```

Comment:
This should reuse the existing logic for finding default implementations instead of rolling its own.

Reply:
More specifically I think these checks could be moved to typechecker::inferDefaultWitnesses().

Reply:
Thank you for pointing this out. I moved the check and it is much cleaner!